src/basic/FStar.Options.fs

```diff
@@ -349,365 +352,404 @@ let mk_spec o : opt =
     let arg =
         match arg with
         | ZeroArgs f ->
-          let g () = set_option' (name, f()) in
+          let g () = set_option name (f()) in
           ZeroArgs g
 
         | OneArg (f, d) ->
-          let g x = set_option' (name, f x) in
+          let g x = set_option name (f x) in
           OneArg (g, d) in
     ns, name, arg, desc
 
-let cons_extract_module s  =
-    List (String.lowercase s::get_extract_module() |> List.map String)
+let accumulated_option name value =
+    let prev_values = Util.dflt [] (lookup_opt name (as_option as_list')) in
+    mk_list (value :: prev_values)
 
-let cons_extract_namespace s  =
-    List (String.lowercase s::get_extract_namespace() |> List.map String)
+let reverse_accumulated_option name value =
+    mk_list ((lookup_opt name as_list') @ [value])
+
+let accumulate_string name post_processor value =
+    set_option name (accumulated_option name (mk_string (post_processor value)))
 
 let add_extract_module s =
-    set_option "extract_module" (cons_extract_module s)
+    accumulate_string "extract_module" String.lowercase s
 
 let add_extract_namespace s =
-    set_option "extract_namespace" (cons_extract_namespace s)
-
-let cons_verify_module s  =
-    List (String.lowercase s::get_verify_module() |> List.map String)
-
-let cons_using_facts_from s =
-    set_option "z3refresh" (Bool true);
-    match get_using_facts_from() with
-    | None -> List [String s]
-    | Some l -> List (List.map String (s :: l))
+    accumulate_string "extract_namespace" String.lowercase s
 
 let add_verify_module s =
-    set_option "verify_module" (cons_verify_module s)
+    accumulate_string "verify_module" String.lowercase s
+
+type opt_type =
+| Const of option_val
+  // --cache_checked_modules
+| IntStr of string (* label *)
+  // --z3rlimit 5
+| BoolStr
+  // --admit_smt_queries true
+| PathStr of string (* label *)
+  // --fstar_home /build/fstar
+| SimpleStr of string (* label *)
+  // --admit_except xyz
+| EnumStr of list<string>
+  // --codegen OCaml
+| OpenEnumStr of list<string> (* values *) * string (* label *)
+  // --debug_level …
+| PostProcessed of ((option_val -> option_val) (* validator *) * opt_type (* elem spec *))
+  // For options like --extract_module that require post-processing or validation
+| Accumulated of opt_type (* elem spec *)
+  // For options like --extract_module that can be repeated (LIFO)
+| ReverseAccumulated of opt_type (* elem spec *)
+  // For options like --include that can be repeated (FIFO)
+| WithSideEffect of ((unit -> unit) * opt_type (* elem spec *))
+  // For options like --version that have side effects
+
+exception InvalidArgument of string // option name
+
+let rec parse_opt_val (opt_name: string) (typ: opt_type) (str_val: string) : option_val =
+  try
+    match typ with
+    | Const c -> c
+    | IntStr _ -> mk_int (int_of_string str_val)
```

Comment:
this will throw if the user passes an invalid integer

Reply:
Indeed. Not sure what to do; this was like that before.

Reply:
I guess I could make a safe version of int_of_string

Reply:
```
let safe_int_of_string x = try Some (int_of_string x) with Invalid_argument _ -> None
```